### PR TITLE
Basic Filter Routing Lookup-table Optimisation

### PR DESCRIPTION
**Pull request by mossblaser**
Work done in collaboration with @neworderofjamie.

This branch adds basic optimisation for the lookup table used to direct arriving packets to the appropriate filter banks. This optimisation step is both limited in scope and also relies on knowledge of the format of routing keys used which is not available at the point of optimisation. Despite these limitations it is important for networks with large ensembles (see motivation below).

We have both attempted to justify to ourselves that the optimiser is correct and have tested the new code with another non-trivial model (Chris' letter-drawing model) where things appeared to continue to work. I've also tested it on some of the example scripts which also continue to work.
# Motivation

These changes are the result of a six hour bug tracking mission by myself and @neworderofjamie for which three hours can be squarely blamed on ST's flood-fill region specification format which is as epicly badass as it is undocumented.

This optimisation is required to allow Lisanne Huurdeman to run her grid cell model on SpiNNaker. This network essentially consists of a single recurrently-connected 47-dimensional ensemble with an enormous number of neurons which must be split across a number of cores. In a "small" case which fails the ensemble has 6,400 neurons. This is split across roughly 8 chips using approximately all 16 cores on each.

Without optimisation, this resulted in tables many tens of kilobytes in size which exhaust the memory of a core causing it to crash in turn causing loading to fail as it checks for chips being successfully loaded. This causes PACMAN to crash in a place which reveals only the flood-fill region it last loaded in the debug messages. Much reverse engineering later and we found the cores which had been loaded and which crashed and diagnosed the memory exhaustion.

After our simple optimisation pass in the best case only one entry is required per chip which connects to the ensemble. Tests on Lisanne's network will take place tomorrow.

A simplified example of this problem (requires 48-node board) is given below:

```
import numpy as np
import nengo

class NullSolver(nengo.decoders.Solver):
    """A decoder solver which simply does nothing.
    """

    def __init__(self):
        self.weights = False

    def __call__(self, A, Y, rng=None):
        N = A.shape[1]
        D = Y.shape[1]
        return np.zeros((N,D)), {}


model = nengo.Network(label="Large ensemble bug test")

with model:
    null_solver = NullSolver()
    # Problem known to kick-in somewhere around 6000 neurons
    a = nengo.Ensemble(12000, 47, label="a")
    nengo.Connection(a, a, solver=null_solver)

import nengo_spinnaker
sim = nengo_spinnaker.Simulator(model)

# All that matters is that it starts
sim.run(1)
```

### CODE REVIEW

**Comment by mundya** [comment]
> relies on knowledge of the format of routing keys used which is not available at the point of optimisation

I almost 99% prepared to guarantee that the top 16-bits will _always_ represent `x` and `y`.  The only place where the bottom 16-bits are arranged differently is (currently) for Jorg's robots, and I suspect that it is unlikely that you'd have two incoming connections for those sharing a filter.  The work I'm currently doing in the `keyspace` branch should mean that you can have some information about the key allocation when doing this optimisation (as in, you'll be able to see how the key is split up).

What I would like to see is this functionality pulled off into a function and some unit tests for that function.  I'll have a play with that later if there's time.

[EDIT:] LGTM, I'd just like tests because when the keyspaces come in this will have to be modified.  And tests are good.

## 

> These changes are the result of a six hour bug tracking mission by myself and @neworderofjamie 

That sounds hellish.  I'm a little concerned by the symptom though:

> Without optimisation, this resulted in tables many tens of kilobytes in size which exhaust the memory of a core causing it to crash in turn causing loading to fail as it checks for chips being successfully loaded. 

Looking in the IO_BUF for the failed core should have shown you a MALLOC failing (and the amount of memory it was trying to reserve)?

There's a [heuristic/hack here](https://github.com/ctn-waterloo/nengo_spinnaker/blob/optimise-filter-routing/nengo_spinnaker/utils/filters.py#L46) for how much space is going to be taken up by the routing entries (at the moment it's a worst case of 5 sub-vertices per pre-vertex), it sounds like it might be worth trying to improve this to push the point of failure back up the tool chain -- so PACMAN has a better idea of whether it needs to keep partitioning.  _Yes, that line of code should have been commented._

File: nengo_spinnaker/utils/filters.py [hunk 1]
```diff
@@ -83,9 +91,37 @@ def _post_prepare_routing(self):
         routings = [edge.prevertex.generate_routing_info(se) for se in
                     edge.subedges]
 
-        self.__filter_keys.extend(
+        filter_keys.extend(
             [FilterRoute(r[0], r[1], i, dmask) for r in routings])
 
+    # We cannot optimise cases where connections from cores on a chip go to
+    # different filters. Find these cases
```
Comment (mundya) [inline nengo_spinnaker/utils/filters.py:98]: This comment confused me.  How would "We cannot optimise cases where ... . Build a map of x, y to sets of Filters, where sets contain 1 element we can perform this optimisation", or similar, sound?

**Comment by mossblaser** [comment]
> That sounds hellish.

Possibly at times though on the whole it was hugely fun :D

>  almost 99% prepared to guarantee that the top 16-bits will always represent x and y. The only place where the bottom 16-bits are arranged differently is (currently) for Jorg's robots, and I suspect that it is unlikely that you'd have two incoming connections for those sharing a filter.

Right, though in the event that one does, does this break (or simply fail to optimise fully)?

> What I would like to see is this functionality pulled off into a function and some unit tests for that function. I'll have a play with that later if there's time.

Awesome. With that said, I'm concious that this problem is (as far as I can tell) exactly equivalent to the standard routing table crunching problem (which shockingly looks like it is being done fairly competently by PACMAN). I'm sure nengo_spinnaker isn't a unique SpiNNaker platform in also needing this optimisation for its software and so I'd very much like to see the function you mention live in PACMAN and share tests.

> [...] And tests are good.

All the justification that needs :)

> Looking in the IO_BUF for the failed core should have shown you a MALLOC failing (and the amount of memory it was trying to reserve)?

This is exactly what happens -- you look in the IOBUF for the core in question and it has a MALLOC error explaining what went wrong in helpful detail -- this is what put us on the right track. (Serious kudos to those involved in making the C libraries here!)

> [...] it might be worth trying to improve this to push the point of failure back up the tool chain -- so PACMAN has a better idea of whether it needs to keep partitioning

Yes, though in this case actually partitioning can only make this particular problem worse (everything connected to this neuron will get at least an additional routing entry.

Incidentally, the unoptimised routing entries consists of vast numbers of duplicates: are you putting one in for every dimension? Additionally, imagine the following case: you have two 2D ensembles, A & B, connected by two connections which compute something on A and only 'connect' to the first and second dimensions of B. This should all be optimised away into one connection (as I'm sure it is) however you get two sets of entries in this lookup table (with different "neuron ID" bits presumably from calling the key allocation function multiple times. This might be worth knowing though it is pretty harmless, especially with this optimisation in place.

**Comment by mundya** [comment]
> With that said, I'm concious that this problem is (as far as I can tell) exactly equivalent to the standard routing table crunching problem (which shockingly looks like it is being done fairly competently by PACMAN). I'm sure nengo_spinnaker isn't a unique SpiNNaker platform in also needing this optimisation for its software and so I'd very much like to see the function you mention live in PACMAN and share tests.

Hopefully the new "we'll have sane interfaces" drive will make this feasible.

> Incidentally, the unoptimised routing entries consists of vast numbers of duplicates: are you putting one in for every dimension?

No, that's why we include the routing key in the filter routing.

> however you get two sets of entries in this lookup table (with different "neuron ID" bits presumably from calling the key allocation function multiple times

Which lookup table?  We spot identical connections on output and merge them to share a connection ID (this is what `utils/connections.py` is there for).  The only reason I didn't put any work into combining things on the receiving side is because it's a humongous amount of work (computationally).

**Comment by mundya** [comment]
This will become irrelevant with #88.  I'll keep this open until then though!

**Comment by mundya** [comment]
Fixed #94.

**Pull request closed without merging**